logs/stream/stream.go

```diff
@@ -37,48 +38,131 @@ import (
 // Google Cloud SQL log data looks like this:
 // - Not correctly ordered (lines from Pub/Sub may arrive in any order)
 // - All lines have a timestamp
-// - First line of a message always has log line number (due to fixed log_line_prefix)
-// - Multi-line messages only have the PID and line number in the first message
+// - Always has the log line number, allowing association of related log lines
+// - Multi-line messages are already combined together
+//   (as of Sept 14, 2021 - see https://cloud.google.com/sql/docs/release-notes#September_14_2021)
+
+const InvalidPid int32 = -1
+
+type LogLineReadiness int
+
+const (
+	LogLineNoAction LogLineReadiness = iota
+	LogLineReady
+	LogLineDiscard
+)
+
+func determineLogLineReadiness(logLine state.LogLine, threshold time.Duration, now time.Time, lastReadyMainLogLinePid int32, lastReadyLogLineWithPrefixPid int32) LogLineReadiness {
+	// The easy case: We have a log line with a log_line_prefix, and only need
+	// to check if we are ready to send or whether there could still be
+	// subsequent lines that will show up within the threshold
+	if logLine.LogLevel != pganalyze_collector.LogLineInformation_UNKNOWN {
+		if now.Sub(logLine.CollectedAt) > threshold || (isAdditionalLineLevel(logLine.LogLevel) && logLine.BackendPid == lastReadyMainLogLinePid) {
+			return LogLineReady
+		}
+
+		return LogLineNoAction
+	}
+
+	// Part of a multi-line log line, where we don't have a log_line_prefix,
+	// but we have a PID, and so can make some assumptions of where this
+	// log line might fit. This is the case for Heroku and syslog receivers.
+	if logLine.BackendPid != 0 {
+		if logLine.BackendPid == lastReadyLogLineWithPrefixPid {
+			return LogLineReady
+		}
+
+		return LogLineNoAction
+	}
+
+	// Part of a multi-line log line, where we don't have a log_line_prefix,
+	// or any other contextual information!
+	//
+	// This is the case for self-managed servers where we tail a log file,
+	// and so we can reasonably assume that subsequent unidentifiable lines
+	// directly belong to the line we saw right before.
+	//
+	// There are edge cases where we discard lines here, since we want to avoid
+	// keeping lines forever that we could never associate to anything.
```

Comment:
I think this is out of scope for this PR, but could self-hosted log output like the following get us in trouble:

```
[1]: backend 2: some long line with
[2]:    newlines so this line is a continuation
[3]: backend 1: another long line with
[4]:    newlines so this line is also a continuation
```

Once we sort this, we'd get

```
[3]: backend 1: another long line with
[1]: backend 2: some long line with
[2]:    newlines so this line is a continuation
[4]:    newlines so this line is also a continuation
```

right? And here we'd lose line [4]? Or am I misunderstanding how this works?